tests/src/test/scala/cats/tests/MonadRecInstancesTests.scala

```diff
@@ -0,0 +1,37 @@
+package cats
+package tests
+
+import cats.data.{OptionT, Xor, XorT}
+
+class MonadRecInstancesTests extends CatsSuite {
+  def tailRecMStackSafety[M[_]](implicit M: MonadRec[M], Eq: Eq[M[Int]]): Unit = {
+    val n = 50000
```

Comment:
@TomasMikula I was thinking of something like what [scalaz has](https://github.com/scalaz/scalaz/blob/series/7.3.x/core/src/main/scala/scalaz/BindRec.scala#L32).

Regarding @adelbertc's comment, I have a little hesitation adding something like that as a law. Mostly because it could potentially take a really long time to run for some types. I'm not writing it off as a possibility, just raising a potential concern.